criu/files-reg.c

```diff
@@ -1571,6 +1577,111 @@ static int get_build_id(const int fd, const struct stat *fd_status,
 	return ret;
 }
 
+static inline void calculate_checksum_iterator_init(int *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter = 0;
+}
+
+static inline void calculate_checksum_iterator_next(int *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter += 1;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter += opts.file_validation_chksm_parameter;
+}
+
+static inline bool calculate_checksum_iterator_stop_condition(int iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY) {
+		return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		if (iter >= opts.file_validation_chksm_parameter)
+			return true;
+		else
+			return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Does the actual work of calculating the CRC32C checksum of "some"
+ * parts of the file. These "some" parts depend on the configuration chosen.
+ * By default, the first 1024 bytes of the file or the entire file, which ever
+ * is smaller, is considered.
+ * Returns true if the checksum could be obtained, else it returns false.
+ */
+static bool calculate_checksum(const int fd, const struct stat *fd_status,
+				u32 *checksum)
+{
+	int bit_count, up_bound, low_bound = 0, i = 0;
+	u32 byte, mask;
+	unsigned char *file_header;
+
+	/* At most, the first 10MB of the file is mapped. */
+	up_bound = min_t(size_t, fd_status->st_size, CHKSM_CHUNK_SIZE);
+	file_header = (unsigned char *) mmap(0, up_bound,
+						PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (file_header == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return false;
+	}
+
+	*checksum = 0xFFFFFFFF;
+	calculate_checksum_iterator_init(&i);
+
+	while (!calculate_checksum_iterator_stop_condition(i) &&
+			(i >= low_bound && i < up_bound)) {
+		byte = file_header[i];
+		calculate_checksum_iterator_next(&i);
+
+		/*
+		 * If the new iterator position is outside the mapped region (Bytes in the range
+		 * low_bound and up_bound) of the file, the current region needs to be unmapped
+		 * and the next portion (At most the next 10MB) of the file needs to be mapped.
+		 */
+		if (i >= up_bound && up_bound != fd_status->st_size) {
+			munmap(file_header, up_bound - low_bound);
+			i -= up_bound;
```

Comment:
Lets assume:
opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST
opts.file_validation_chksm_parameter == CHKSM_CHUNK_SIZE+5
and st_size > CHKSM_CHUNK_SIZE*2

In calculate_checksum you would cycle in while() loop untill i == CHKSM_CHUNK_SIZE-1, then calculate_checksum_iterator_next will increase i to CHKSM_CHUNK_SIZE, then you get to "if (i >= up_bound && up_bound != fd_status->st_size) {" true branche, and do "i -= up_bound", so that i == 0 now. Also you set low_bount = up_bound == CHKSM_CHUNK_SIZE. So while loop will stop on the next iteration because i(0) < low_bound(CHKSM_CHUNK_SIZE). You've only checked CHKSM_CHUNK_SIZE bytes, not what is expected CHKSM_CHUNK_SIZE+5.

Looks like the logic is broken.

Reply:
Sorry about that, I had overlooked it. I've fixed it and pushed the new code.

Reply:
Maybe we can add a test for this case?

Reply:
Yeah I was planning on adding a test along these lines along with 2 others:
- `opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD` and`opts.file_validation_chksm_parameter == CHKSM_CHUNK_SIZE+10`.
- Alter the file after dump but before restore, this should cause restore to fail. If restore succeeds then the test fails.

These 3 tests should be enough for the checksum method?

Reply:
> These 3 tests should be enough for the checksum method?

I would say we need to add at least one test for each checksum method (3 tests). And for tests with chksm-parameter we should likely have two tests 1)if parameter < CHKSM_CHUNK_SIZE and 2) > CHKSM_CHUNK_SIZE. So 5 tests would be nice.

For the exact case you've fixed we should alter e.g. the last byte of "the first 'N'' to check that we detect broken file even if only the 'N'-th byte is changed.

To sum-up: more tests are always better, but don't spend too much time on it =) Add some tests to PR and we will see if it's enough.